src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java

```diff
@@ -0,0 +1,107 @@
+package org.springframework.data.elasticsearch.core;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.common.collect.MapBuilder;
+import org.elasticsearch.index.query.MoreLikeThisQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.springframework.data.elasticsearch.annotations.Document;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Artur Konczak
+ * @author Mohsin Husen
+ * @author Mark Paluch
+ * @author Mark Janssen
+ * @author Sascha Woo
+ */
+public abstract class AbstractElasticsearchTemplate implements ElasticsearchOperations {
+
+	protected static String[] toArray(List<String> values) {
+
+		Assert.notNull(values, "values must not be null");
+		return values.toArray(new String[values.size()]);
+	}
+
+	protected static MoreLikeThisQueryBuilder.Item[] toArray(MoreLikeThisQueryBuilder.Item... values) {
+		return values;
+	}
+
+	@Override
+	public ElasticsearchPersistentEntity<?> getPersistentEntityFor(Class<?> type) {
+
+		Assert.notNull(type, "type must not be null!");
+		Assert.isTrue(type.isAnnotationPresent(Document.class), "Unable to identify index name. " + type.getSimpleName()
+				+ " is not a Document. Make sure the document class is annotated with @Document(indexName=\"foo\")");
+
+		return getElasticsearchConverter().getMappingContext().getRequiredPersistentEntity(type);
+	}
+
+	protected List<String> defaultIfEmpty(List<String> strs, Supplier<String> defaultString) {
+		return CollectionUtils.isEmpty(strs) ? Arrays.asList(defaultString.get()) : strs;
+	}
+
+	protected String defaultIfEmpty(String str, Supplier<String> defaultString) {
+		return StringUtils.isEmpty(str) ? defaultString.get() : str;
+	}
+
+	protected List<String> extractIds(SearchResponse response) {
```

Comment:
We should find a better place for this utility method.

Reply:
For now, I would leave it here and later move this code to the `ElasticsearchRestTemplate`, where it is currently used only once.